src/sagemaker_containers/encoder.py

```diff
@@ -0,0 +1,175 @@
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import json
+import textwrap
+
+import numpy as np
+from six import BytesIO, StringIO
+
+from sagemaker_containers import content_types
+
+
+class NpyEncoder(object):
+    content_type = content_types.NPY
+
+    @staticmethod
+    def encode(np_array):  # type: (np.array) -> object
+        """Encode an numpy array into memory.
+
+        Args:
+            np_array (np.array): np array object to be serialized.
+
+        Returns:
+            (obj): serialized np array.
+        """
+        buffer = BytesIO()
+        np.save(buffer, np_array)
+        return buffer.getvalue()
+
+
+class NpyDecoder(object):
+    content_type = content_types.NPY
+
+    @staticmethod
+    def decode(obj):  # type: (object) -> np.array
+        """Decode arrays or pickled objects from memory into numpy arrays.
+
+        Args:
+            obj (object): object to be deserialized.
+
+        Returns:
+            (np.array): deserialized numpy array.
+        """
+        stream = BytesIO(obj)
+        return np.load(stream)
+
+
+class JsonEncoder(json.JSONEncoder):
+    content_type = content_types.JSON
+
+    def default(self, o):
+        if hasattr(o, 'read'):
+            return o.read()
+
+        if hasattr(o, 'tolist'):
+            return o.tolist()
+
+        return super(JsonEncoder, self).default(o)
+
+
+class JsonDecoder(object):
+    content_type = content_types.JSON
+
+    @staticmethod
+    def decode(string):  # type: (object) -> object
+        """Decode a JSON string in a Python object.
+
+        Args:
+            string (str): JSON string to be decoded.
+
+        Returns:
+            (object): deserialized Python object.
+        """
+        return json.loads(string)
+
+
+class CsvDecoder(object):
+    content_type = content_types.CSV
+
+    @staticmethod
+    def decode(string):  # type: (str) -> object
+        """Decode a CSV string in a Python object.
+
+        Args:
+            string (str): CSV string to be decoded
+
+        Returns:
+            (object): decoded Python object.
+        """
+        stream = StringIO(string)
+        return np.genfromtxt(stream, dtype=np.float32, delimiter=',')
+
+
+class CsvEncoder(object):
+    content_type = content_types.CSV
+
+    @staticmethod
+    def encode(obj):  # type: (object) -> str
+        """Encode python objects, streams, numpy arrays into CSV.
+
+        Args:
+            obj (object): object to be encoded.
+
+        Returns:
+            str:  decoded object.
+        """
+        stream = StringIO()
+        np.savetxt(stream, obj, delimiter=',', fmt='%s')
+        return stream.getvalue()
+
+
+DEFAULT_ENCODERS = [JsonEncoder(), CsvEncoder(), NpyEncoder()]
```

Comment:
ok, it can be a list, but then DefaultDecoder should turn it into a dict in its constructor. also, allowing people to pass in a list doesn't mean we need to expose our list as a (mutable!) module variable. it could just be default in DefaultDecoder's __init__ if no list is supplied.